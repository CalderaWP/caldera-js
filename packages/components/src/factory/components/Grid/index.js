export { Rows } from './Rows';
export { Row } from './Row';
export { Column } from './Column';
