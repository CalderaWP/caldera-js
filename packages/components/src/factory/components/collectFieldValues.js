/**
 * Get the values of all config fields
 * @param fields
 */
export const collectFieldValues = fields => {
	const values = {};
	fields.forEach(field => {
		if ('object' === typeof field) {
			values[field.fieldId] = field.value
				? field.value
				: null;
		}
	});
	return values;
};
