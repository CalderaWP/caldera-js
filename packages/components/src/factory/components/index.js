export { Field } from './Field';
export { FieldArea } from './FieldArea';
export { Fields } from './Fields';
export { Rows } from './Grid/Rows';
export { Row } from './Grid/Row';
export { Column } from './Grid/Column';
