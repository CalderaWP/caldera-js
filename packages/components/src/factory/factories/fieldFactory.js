import {
	RemotePost,
	TextAreaField,
	ToggleField,
	RadioField,
	FieldWrapper,
	FieldSet,
	SelectField,
	InputField,
	RichText,
	MagicRichText,
	isValidHtml5type,
	AutoCompleteField,
	FormFieldsAutoComplete
} from '@calderawp/components';
import React from 'react';

/**
 * Create a field from an object describing it.
 *
 * Note: field.render is a render prop. If it is passed, field is passed to it.
 *
 * @param field
 * @param onChange
 * @param onBlur
 * @return {*}
 */
export const fieldFactory = (field, onChange, onBlur) => {
	const {
		fieldType,
		label,
		attributes,
		options,
		fieldId,
		messages,
		render
	} = field;

	if (render) {
		return React.createElement(render, field);
	}

	switch (fieldType) {
		case 'checkboxes':
			return (
				<FieldSet
					fieldType={fieldType}
					legend={label}
					attributes={attributes}
				>
					{options.map(option => {
						const {
							value,
							label,
							description,
							attributes
						} = option;
						const optionId = option.hasOwnProperty('id')
							? option.id
							: `opt-${fieldId}-${option.value}`;
						return (
							<InputField
								key={optionId}
								id={optionId}
								fieldId={optionId}
								value={value}
								label={label}
								description={description}
								html5type={'checkbox'}
								attributes={attributes}
								onChange={onChange}
							/>
						);
					})}
				</FieldSet>
			);
		case 'magic-richtext' :
			return <MagicRichText {...field} onChange={onChange}/>;
		case 'autocomplete' :
			return <AutoCompleteField {...field} onChange={onChange}/>;
		case'richtext' :
			return <RichText {...field} onChange={onChange}/>;
			case'fields-autocomplete' :
			return <FormFieldsAutoComplete {...field} onChange={onChange}/>;
		case 'textarea':
			return <TextAreaField {...field} onChange={onChange}/>;
		case 'toggle':
			return <ToggleField {...field} onChange={onChange}/>;
		case 'radio':
			return <RadioField {...field} onChange={onChange}/>;
		case 'select':
		case 'dropdown':
			return <SelectField {...field} onChange={onChange}/>;
		case 'text':
		case 'email':
		case 'number':
		case 'input':
		default:
			if (isValidHtml5type(fieldType)) {
				if (field.html5type !== fieldType) {
					field = {...field, html5type: fieldType};
				}
			} else {
				field.html5type = 'text';
			}
			return <InputField {...field} onChange={onChange}/>;
	}
};
