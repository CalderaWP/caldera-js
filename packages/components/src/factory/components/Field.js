import React, { Fragment } from 'react';
import { fieldFactory } from '../factories/fieldFactory';

export const Field = ({ field, onChange, onBlur, messages }) => (
	<Fragment>{fieldFactory(field, onChange, onBlur, messages)}</Fragment>
);
