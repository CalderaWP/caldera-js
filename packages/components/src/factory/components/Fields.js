import React, { Fragment } from 'react';
import { Field } from './Field';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { collectFieldValues } from './collectFieldValues';
export const Fields = ({ fields, onChange, onBlur, className }) => (
	<div className={classNames('caldera-fields-wrapper', className)}>
		{fields.map(field => (
			<Field
				key={field.fieldId}
				field={field}
				onChange={newValue => {
					onChange({
						...collectFieldValues(fields),
						[field.fieldId]: newValue
					});
				}}
				onBlur={onBlur}
			/>
		))}
	</div>
);

Fields.propTypes = {
	field: PropTypes.shape({
		fieldId: PropTypes.string.isRequired
		//input propTypes?
	}),
	onChange: PropTypes.func.isRequired,
	onBlur: PropTypes.func,
	className: PropTypes.string
};
