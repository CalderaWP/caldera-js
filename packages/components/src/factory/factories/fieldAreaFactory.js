import { FieldWrapper, Message } from '@calderawp/components';
import React, { Fragment } from 'react';
import classNames from 'classnames';
import { fieldFactory } from './fieldFactory';

export const fieldAreaFactory = (
	field,
	onChange,
	onBlur,
	fieldErrors,
	fieldsTouch
) => {
	const { fieldType, fieldId, required } = field;
	const error = fieldErrors && fieldErrors[fieldId];
	const touched = fieldsTouch && fieldsTouch[fieldId];
	return (
		<FieldWrapper
			fieldType={'text'}
			className={classNames('caldera-field-group', {
				'has-error': touched && error,
				'is-required': required
			})}
		>
			<Fragment key={`${fieldId}-1`}>
				{fieldFactory(field, onChange, onBlur)}
			</Fragment>
			{touched && error && (
				<Fragment key={`${fieldId}-2`}>
					<Message
						message={{
							error: true,
							message: error
						}}
					/>
				</Fragment>
			)}
		</FieldWrapper>
	);
};
