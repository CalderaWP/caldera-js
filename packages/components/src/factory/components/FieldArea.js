import React, { Fragment } from 'react';
import { fieldAreaFactory } from '../factories/fieldAreaFactory';

export const FieldArea = ({
	field,
	onChange,
	onBlur,
	fieldErrors,
	fieldsTouch,
}) => (
	<Fragment>
		{fieldAreaFactory(field, onChange, onBlur, fieldErrors, fieldsTouch)}
	</Fragment>
);
