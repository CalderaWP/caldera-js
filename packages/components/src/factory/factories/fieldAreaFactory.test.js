import { fieldAreaFactory } from './fieldAreaFactory';
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField, textAreaField, toggleField, autoCompleteField
} from '../fields.fixtures';

import { fieldWrapperClassNames } from '@calderawp/components';
import { fieldFactory } from './fieldFactory';

describe('fieldAreaFactory', () => {
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});
	it('Creates a text field', () => {
		const component = renderer.create(
			fieldAreaFactory(textField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates an select field', () => {
		const component = renderer.create(
			fieldAreaFactory(selectField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('should create textarea', () => {
		const component = renderer.create(
			fieldAreaFactory(textAreaField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('should create toggle', () => {
		const component = renderer.create(
			fieldAreaFactory(toggleField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Should create an autocomplete field', () => {
		const component = renderer.create(
			fieldAreaFactory(autoCompleteField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
});

describe('change handlers', () => {
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});

	it('Changes calls change handler', () => {
		const component = mount(fieldAreaFactory(textField, onChange, onBlur));
		component.find('input').simulate('change');
		expect(onChange.mock.calls.length).toBe(1);
	});
});

describe('Showing errors', () => {
	//fieldErrors, fieldsTouch
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});
	it('Does not display error class on field wrapper if field has not been touched', () => {
		const errorMessage = 'Fails!';
		const fieldErrors = {
			[textField.fieldId]: errorMessage
		};
		const component = mount(
			fieldAreaFactory(textField, onChange, onBlur, fieldErrors, {})
		);
		expect(
			component.find('.caldera-field-group').hasClass('has-error')
		).toBe(false);
	});
	it('Does not display error class on field wrapper if field has  been touched', () => {
		const errorMessage = 'Fails!';
		const fieldErrors = {
			[textField.fieldId]: errorMessage
		};
		const fieldTouched = {
			[textField.fieldId]: true
		};
		const component = mount(
			fieldAreaFactory(
				textField,
				onChange,
				onBlur,
				fieldErrors,
				fieldTouched
			)
		);
		expect(
			component
				.find('.caldera-field-group')
				.first()
				.hasClass('has-error')
		).toBe(true);
	});

	it('Does not display Message component if field has not been touched', () => {
		const errorMessage = 'Fails!';
		const fieldErrors = {
			[textField.fieldId]: errorMessage
		};
		const fieldTouched = {};
		const component = mount(
			fieldAreaFactory(
				textField,
				onChange,
				onBlur,
				fieldErrors,
				fieldTouched
			)
		);
		expect(component.find('.caldera-components-error').length).toBe(0);
	});

	it('Does display Message component if field has been touched', () => {
		const errorMessage = 'Fails!';
		const fieldErrors = {
			[textField.fieldId]: errorMessage
		};
		const fieldTouched = {
			[textField.fieldId]: true
		};
		const component = mount(
			fieldAreaFactory(
				textField,
				onChange,
				onBlur,
				fieldErrors,
				fieldTouched
			)
		);

		expect(
			component.find('.caldera-components-error').hasClass('has-error')
		).toBe(true);
		expect(component.find('.caldera-components-error').text()).toBe(
			errorMessage
		);
	});

	it('Does not display Message component if field has no errors field has been touched', () => {
		const errorMessage = 'Fails!';
		const fieldErrors = {
			[textField.fieldId]: false
		};
		const fieldTouched = {
			[textField.fieldId]: true
		};
		const component = mount(
			fieldAreaFactory(
				textField,
				onChange,
				onBlur,
				fieldErrors,
				fieldTouched
			)
		);
		expect(component.find('.caldera-components-error').length).toBe(0);
	});

	test('Field can be a component using render prop', () => {
		const _Field = ({ fieldId }) => (
			<input id={fieldId} type={'number'} key={88} />
		);
		const field = {
			...selectField,
			render: _Field
		};
		const component = mount(
			fieldAreaFactory(field, onChange, onBlur, {}, {})
		);
		expect(component.find('#selectFieldId').length).toEqual(1);
		expect(component.find('#selectFieldId').prop('type')).toEqual('number');
	});
});
