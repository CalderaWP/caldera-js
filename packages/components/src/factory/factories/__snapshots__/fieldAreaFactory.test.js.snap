// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldAreaFactory Creates a text field 1`] = `
<div
  className="caldera-field-wrapper caldera-field-wrapper-text"
>
  <div
    className="components-base-control caldera-field caldera-field-text"
  >
    <div
      className="components-base-control__field"
    >
      <label
        className="components-base-control__label"
        htmlFor="inspector-text-control-0"
      >
        First Name
      </label>
      <input
        aria-describedby="inspector-text-control-0__help"
        className="components-text-control__input"
        id="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder={undefined}
        type="text"
        value="roy"
      />
    </div>
    <p
      className="components-base-control__help"
      id="inspector-text-control-0__help"
    >
      your first name
    </p>
  </div>
</div>
`;

exports[`fieldAreaFactory Creates an select field 1`] = `
<div
  className="caldera-field-wrapper caldera-field-wrapper-text"
>
  <div
    className="components-base-control caldera-field caldera-field-select"
  >
    <div
      className="components-base-control__field"
    >
      <label
        className="components-base-control__label"
        htmlFor="inspector-select-control-0"
      >
        Select Field Label
      </label>
      <select
        aria-describedby="inspector-select-control-0__help"
        className="components-select-control__input"
        id="selectFieldId"
        multiple={false}
        onChange={[Function]}
        value={1}
      >
        <option
          value={0}
        >
          Zero
        </option>
        <option
          value={1}
        >
          One
        </option>
        <option
          value={2}
        >
          Two
        </option>
      </select>
    </div>
    <p
      className="components-base-control__help"
      id="inspector-select-control-0__help"
    >
      Select field description
    </p>
  </div>
</div>
`;

exports[`fieldAreaFactory Should create an autocomplete field 1`] = `
<div
  className="caldera-field-wrapper caldera-field-wrapper-text"
>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <label
        className="components-base-control__label"
        htmlFor="autoComplete1"
      >
        Options
      </label>
      <div
        className="components-form-token-field"
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        tabIndex="-1"
      >
        <label
          className="components-form-token-field__label"
          htmlFor="components-form-token-input-0"
        >
          Add item
        </label>
        <div
          className="caldera-field caldera-field-autocomplete components-form-token-field__input-container"
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          tabIndex="-1"
        >
          <span
            className="components-form-token-field__token"
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            title={undefined}
          >
            <span
              className="components-form-token-field__token-text"
              id="components-form-token-field__token-text-0"
            >
              <span
                className="screen-reader-text"
              >
                One (1 of 2)
              </span>
              <span
                aria-hidden="true"
              >
                One
              </span>
            </span>
            <button
              aria-describedby="components-form-token-field__token-text-0"
              aria-label="Remove item"
              className="components-button components-icon-button components-form-token-field__remove-token"
              disabled={undefined}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="button"
            >
              <svg
                aria-hidden="true"
                className="dashicon dashicons-dismiss"
                focusable="false"
                height={20}
                role="img"
                viewBox="0 0 20 20"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
                />
              </svg>
            </button>
          </span>
          <span
            className="components-form-token-field__token"
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            title={undefined}
          >
            <span
              className="components-form-token-field__token-text"
              id="components-form-token-field__token-text-1"
            >
              <span
                className="screen-reader-text"
              >
                Three (2 of 2)
              </span>
              <span
                aria-hidden="true"
              >
                Three
              </span>
            </span>
            <button
              aria-describedby="components-form-token-field__token-text-1"
              aria-label="Remove item"
              className="components-button components-icon-button components-form-token-field__remove-token"
              disabled={undefined}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="button"
            >
              <svg
                aria-hidden="true"
                className="dashicon dashicons-dismiss"
                focusable="false"
                height={20}
                role="img"
                viewBox="0 0 20 20"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
                />
              </svg>
            </button>
          </span>
          <input
            aria-activedescendant={undefined}
            aria-autocomplete="list"
            aria-describedby="components-form-token-suggestions-howto-0"
            aria-expanded={false}
            aria-owns={undefined}
            autoCapitalize={undefined}
            autoComplete={undefined}
            className="components-form-token-field__input"
            disabled={false}
            id="components-form-token-input-0"
            onBlur={[Function]}
            onChange={[Function]}
            role="combobox"
            size={1}
            type="text"
            value=""
          />
        </div>
        <div
          className="screen-reader-text"
          id="components-form-token-suggestions-howto-0"
        >
          Separate with commas
        </div>
      </div>
    </div>
    <p
      className="components-base-control__help"
      id="autoComplete1__help"
    >
      Help text
    </p>
  </div>
</div>
`;

exports[`fieldAreaFactory should create textarea 1`] = `
<div
  className="caldera-field-wrapper caldera-field-wrapper-text"
>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <label
        className="components-base-control__label"
        htmlFor="inspector-textarea-control-0"
      >
        Your message
      </label>
      <textarea
        aria-describedby="inspector-textarea-control-0__help"
        className="components-textarea-control__input"
        cols={8}
        id="textArea1"
        maxLength={500}
        minlength={10}
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        rows={5}
        value="The content of text area"
      />
    </div>
    <p
      className="components-base-control__help"
      id="inspector-textarea-control-0__help"
    >
      your first name
    </p>
  </div>
</div>
`;

exports[`fieldAreaFactory should create toggle 1`] = `
<div
  className="caldera-field-wrapper caldera-field-wrapper-text"
>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <label
        className="components-base-control__label"
        htmlFor="toggleFieldLabel"
      >
        Toggle Label
      </label>
      <div
        className="components-base-control components-toggle-control"
      >
        <div
          className="components-base-control__field"
        >
          <span
            className="components-form-toggle is-checked"
          >
            <input
              aria-describedby={undefined}
              checked={true}
              className="components-form-toggle__input"
              id="inspector-toggle-control-0"
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="components-form-toggle__track"
            />
            <span
              className="components-form-toggle__thumb"
            />
            <svg
              aria-hidden="true"
              className="components-form-toggle__on"
              focusable="false"
              height="6"
              role="img"
              viewBox="0 0 2 6"
              width="2"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h2v6H0z"
              />
            </svg>
          </span>
          <label
            className="components-toggle-control__label"
            htmlFor="inspector-toggle-control-0"
          />
        </div>
      </div>
    </div>
    <p
      className="components-base-control__help"
      id="toggleFieldLabel__help"
    >
      Toggle field description
    </p>
  </div>
</div>
`;
